src/ellipse2d.jl

```diff
@@ -1,30 +1,15 @@
-"""
-```
-img_with_ellipse = ellipse(img, center, radiusy, radiusx)
-img_with_ellipse = ellipse(img, center, color, radiusy, radiusx)
-img_with_ellipse = ellipse(img, y, x, radiusy, radiusx)
-img_with_ellipse = ellipse(img, y, x, color, radiusy, radiusx)
-```
-Draws a ellipse on the input image given the `center` as a `CartesianIndex{2}` or
-as coordinates `(y, x)` using the specified `color`. If `color` is not specified,
-`one(eltype(img))` is used.
-"""
-ellipse{T<:Colorant}(img::AbstractArray{T, 2}, args...) = ellipse!(copy(img), args...)
+#Ellipse methods
 
-function ellipse!{T<:Colorant}(img::AbstractArray{T, 2}, center::CartesianIndex{2}, color::T, radiusy::Real, radiusx::Real)
-    ellipse!(img, center[1], center[2], color, radiusy, radiusx)
-end
-
-function ellipse!{T<:Colorant}(img::AbstractArray{T, 2}, y::Int, x::Int, radiusy::Real, radiusx::Real)
-	ellipse!(img, y, x, one(T), radiusy, radiusx)
-end
+Ellipse(x::Int, y::Int, ρx::Real, ρy::Real) = Ellipse(Point(x,y), ρx, ρy)
+Ellipse(p::CartesianIndex{2}, ρx::Real, ρy::Real) = Ellipse(Point(p), ρx, ρy)
+Ellipse(circle::CirclePointRadius) = Ellipse(circle.center, circle.ρ, circle.ρ)
 
-function ellipse!{T<:Colorant}(img::AbstractArray{T, 2}, y::Int, x::Int, color::T, radiusy::Real, radiusx::Real)
+function draw!{T<:Colorant}(img::AbstractArray{T, 2}, ellipse::Ellipse, color::T)
 	ys = Int[]
 	xs = Int[]
-	for i in y : y + radiusy
-		for j in x : x + radiusx
-			val = ((i - y) / radiusy) ^ 2 + ((j - x) / radiusx) ^ 2
+	for i in ellipse.center.y : ellipse.center.y + ellipse.ρy
+		for j in ellipse.center.x : ellipse.center.x + ellipse.ρx
+			val = ((i - ellipse.center.y) / ellipse.ρy) ^ 2 + ((j - ellipse.center.x) / ellipse.ρx) ^ 2
```

Comment:
Concerning performance bottleneck and parameteric types (please @timholy correct me here if I am mistaken):

Consider this nested loop. If the member variables of the variable `ellipse` are of abstract types, then two things are true. 

- The member variables of `ellipse` are located on the heap, not the stack, regardless if the type `Ellipse` itself is `immutable`. 

- The loop variables `i` and `j` are not inferrable, which means that every function in the inner loop that uses those is invoked using dynamic dispatch (runtime method lookup). Dynamic dispatch itself isn't a problem per say (it is a useful option to have), but it is undesirable to have dynamic dispatch happen inside a (nested) loop, because that code is potentially executed many times.